pkg/bu_server/trade_document/file_based_ebl.go

```diff
@@ -298,6 +309,83 @@ func (c *_FileBaseEBLController) List(ctx context.Context, req ListFileBasedEBLR
 	return res, nil
 }
 
+func (c *_FileBaseEBLController) Transfer(ctx context.Context, ts int64, req TransferEBLRequest) (bill_of_lading.BillOfLadingPack, error) {
+	currentTime := model.NewDateTimeFromUnix(ts)
+	if err := ValidateTransferEBLRequest(req); err != nil {
+		return bill_of_lading.BillOfLadingPack{}, err
+	}
+
+	if err := c.checkBUExistence(ctx, req.Application, []string{req.TransferBy}); err != nil {
+		return bill_of_lading.BillOfLadingPack{}, err
+	}
+
+	tx, err := c.storage.CreateTx(ctx, storage.TxOptionWithWrite(true), storage.TxOptionWithIsolationLevel(sql.LevelSerializable))
+	if err != nil {
+		return bill_of_lading.BillOfLadingPack{}, err
+	}
+	defer func() { _ = tx.Rollback(ctx) }()
+
+	oldPack, oldHash, err := c.getEBL(ctx, tx, req.ID)
+	if err != nil {
+		return bill_of_lading.BillOfLadingPack{}, err
+	}
+	if err = IsFileEBLTransferable(&oldPack, req.TransferBy, true); err != nil {
+		return bill_of_lading.BillOfLadingPack{}, err
+	}
+
+	var bl *bill_of_lading.BillOfLading
+	for i := len(oldPack.Events) - 1; i >= 0; i-- {
+		if oldPack.Events[i].BillOfLading != nil {
+			bl = oldPack.Events[i].BillOfLading
+			break
+		}
+	}
+	if bl == nil {
+		return bill_of_lading.BillOfLadingPack{}, errors.New("no bill of lading found in the pack")
+	}
+
+	parties := lo.Map(bl.BillOfLading.ShippingInstruction.DocumentParties, func(p bill_of_lading.DocumentParty, _ int) string {
+		return p.Party.IdentifyingCodes[0].PartyCode
+	})
+
+	_, currentOwnerIdx, _ := lo.FindIndexOf(parties, func(p string) bool { return p == req.TransferBy })
+
+	blPack := bill_of_lading.BillOfLadingPack{
+		ID:           oldPack.ID,
+		Version:      oldPack.Version + 1,
+		ParentHash:   oldHash,
+		Events:       oldPack.Events,
+		CurrentOwner: parties[currentOwnerIdx+1],
```

Comment:
This implementation works only when the parties are in the order of "Issuer", "Shipper", "Consignee", "ReleaseAgent". It's not safe. We should identify the role of the caller first then decide which party is the next owner based its role.